modules/http-server/src/test/java/org/glassfish/grizzly/http/server/NetworkListenerTest.java

```diff
@@ -241,7 +243,14 @@ public void testGracefulShutdownGracePeriodExpired() throws Exception {
             Future<HttpServer> gracefulFuture = server.shutdown(1, TimeUnit.SECONDS);
 
             final BufferedReader reader = new BufferedReader(new InputStreamReader(c.getInputStream(), Charsets.UTF8_CHARSET));
-            final String content = reader.readLine();
+            String content;
+            try {
+                // null is expected here as the grace period will expire
```

Comment:
This comment doesn't make much sense. Is `null` expected on this line? Wouldn't it result in a nullptr exception?
Or it's expected that reader.readLine() throws IOException and then it's OK to set the content to `null`?

I think what you want to say is that `content == null` means the grace period expired. And on newer JDKs the behavior changes - an exception is thrown if grace period expires, instead of returning null.

What about this comment:

```suggestion
                // this should result in null in older JDKs or IOException on newer JDKs as the grace period expires 
```

Reply:
Perhaps I should add a little more explanation. I understand why you were confused. I checked the actual JDK source and noticed that the behavior changed (JDK17 vs. JDK25). However, I didn't find any notices or changelogs regarding this behavior change. However, I was only aware that security and network protocols have become more stringent since JDK 25.

I will try to provide more detailed explanations and supplement the content later.

Reply:
> Perhaps I should add a little more explanation. I understand why you were confused. I checked the actual JDK source and noticed that the behavior changed (JDK17 vs. JDK25). 

Maybe this https://bugs.openjdk.org/browse/JDK-8335135 - https://github.com/openjdk/jdk/commit/b6ffb442acb4a222f017868433eff213d9b84ed8

Reply:
@OndroMih The `testGracefulShutdownGracePeriodExpired` test does not completely test whether the listener is gracefully shut down.
Graceful shutdown means shutting down when there are no currently active or in-progress http requests and all processing has been completed.

This test generates a shutdown event, delivers it to each listener and registered filter, maintains an unfinished state, and assumes that forced shutdown occurs successfully.

Looking at the test case, Grizzly generates a shutdown event and has a grace period of 1 second.
Grizzly handler first sends a response containing http status and header, waits 2 seconds (expects grace period to expire), and then writes msg.
The core of the test appears to be testing whether or not the shutdown proceeds after the grace period expires after the shutdown event is delivered/propagated through whether or not this msg is received.
In other words, we are verifying that after expiration, shutdown occurs and no msg is received. It seems that the client side probably assumed an EOF situation.

BufferedReader#readLine() is explained as follows:
```
Returns:
A String containing the contents of the line, not including any line-termination characters, or null if the end of the stream has been reached without reading any characters
```
Additionally, there was no change to the intent of returning null in EOF situations.
https://github.com/openjdk/jdk/blob/2c07214d7c075da5dd4a4e872aef29f58cef2bae/src/java.base/share/classes/java/io/BufferedReader.java#L332-L338

However, like the part linked by @pzygielo above, the fill() behavior has been changed to throw an IOException at EOF before returning null.

In a situation where null is returned in the old JDK and an IOException occurs in the new JDK, I briefly thought about changing the test case in a different way to prevent such changes in the JDK, but for comparison purposes, I tried to maintain the existing source as much as possible, so it ended up like this.
I thought it would be less confusing to leave at least a hint rather than just applying the modifications, and I thought it was a test case without much meaning, so I made comments too roughly, which may have caused more confusion.

As you advised, I will submit an edit to leave a more clear comment. :)

Reply:
Thanks @carryel, makes perfect sense now! I was maybe a little too pedantic. i just thought it’s good to leave a clear comment in the code for people that come across it in the future :)